src/api/exchange/ExchangeApi.ts

```diff
@@ -101,21 +115,175 @@ export interface Order {
   remainingAmount: BN
 }
 
+/**
+ * BaseTradeEvent uses only info available on the emitted Trade event
+ */
+export interface BaseTradeEvent {
+  // order related
+  orderId: string
+  sellTokenId: number
+  buyTokenId: number
+  sellAmount: BN
+  buyAmount: BN
+  // block related
+  txHash: string
+  txIndex: number
+  blockNumber: number
+  id: string // txHash + | + txIndex
+}
+
+/**
+ * TradeEvent enriches BaseTradeEvent with block, order and token data
+ */
+export interface TradeEvent extends BaseTradeEvent {
+  batchId: number
+  hashKey: string // orderId + batchId, to find reverts
+  indexOnBatch: number // tracks trade position on batch, in case of reverts
+  time: Date
+  buyToken: TokenDetails
+  sellToken: TokenDetails
+  limitPrice: BigNumber
+  fillPrice: BigNumber
+  reverted: boolean
+}
+
 export interface GetOrdersPaginatedResult {
   orders: AuctionElement[]
   nextIndex?: number
 }
 
+// TODO: move to const/config
+const CONTRACT_DEPLOYMENT_BLOCK = {
+  1: 9340147,
+  4: 5844678,
+}
+
+// Syntactic sugar. Aliasing event types to make code cleaner
+type TradeSubscription = ReturnType<BatchExchangeContract['events']['Trade']>
+// type TradeEventData = TradeSubscription extends Subscription<infer U> ? U : never
+
+interface Subscriptions {
+  trade: { [networkId: number]: TradeSubscription }
+}
 /**
  * Basic implementation of Stable Coin Converter API
  */
 export class ExchangeApiImpl extends DepositApiImpl implements ExchangeApi {
+  private subscriptions: Subscriptions = { trade: {} }
+
   public constructor(injectedDependencies: Params) {
     super(injectedDependencies)
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     ;(window as any).exchange = this._contractPrototype
   }
 
+  public async getPastTrades({ userAddress, networkId }: PastEventsParams): Promise<BaseTradeEvent[]> {
+    const contract = await this._getContract(networkId)
+
+    // to get all past events
+    const tradeEvents = await contract.getPastEvents('Trade', {
+      fromBlock: CONTRACT_DEPLOYMENT_BLOCK[networkId],
+      filter: { owner: userAddress },
+    })
+
+    logDebug(
+      `[ExchangeApiImpl] Fetched ${tradeEvents.length} trades for address ${userAddress} on network ${networkId}`,
+    )
+
+    return tradeEvents.filter(event => !event['removed']).map(this.parseTradeEvent)
+  }
+
+  public async subscribeToTradeEvent(params: SubscriptionParams): Promise<() => void> {
+    const { userAddress, networkId, callback } = params
+
+    // Remove any active subscription
+    this.unsubscribeToTradeEvent()
+
+    logDebug(`[ExchangeApiImpl] subscribing to trade events for address ${userAddress} and networkId ${networkId}`)
+
+    const subscription = await this.getTradeSubscription(params)
+    // Two ways of subscribing
+
+    // 1: with error handling
+    subscription.subscribe((error, event) => {
```

Comment:
Are we using subscription? 
I thought one issue with that according to @Velenir  is that some wallets don't allow to subscribe.
If I recall from the sprint planning, The alternative was using a default provider, but that was a bit less scalable.

This is to keep up with the new trade events?
The idea we said it might be worth trying was, to use getPastLogs with a range of logs, and only do that if the bloom filter says so. 

For some reason, we trust more in `getPastLogs` than subscribe. There's testimonies of missed events?? Also, are we sure our strategy on reorgs?

Reply:
> Are we using subscription?
I thought one issue with that according to @Velenir is that some wallets don't allow to subscribe.
If I recall from the sprint planning, The alternative was using a default provider, but that was a bit less scalable.

Yes, we talked about that and to be honest I did not remember about it until now :)
Did not get to test it with any mobile wallet though.

> This is to keep up with the new trade events?

Yep.

> The idea we said it might be worth trying was, to use getPastLogs with a range of logs, and only do that if the bloom filter says so.

I thought the bloom filter was for getting balances.

> For some reason, we trust more in getPastLogs than subscribe. There's testimonies of missed events?? Also, are we sure our strategy on reorgs?

I can't testify in favor nor against any approach.

What I can tell is that, while testing, I'm somehow subscribing more than once to the events, or at least getting more than one notification per event. Tried to always unsubscribe before new subscriptions but that still did not help.

---

In summary, you are saying that we might just not have this method, and instead rely on `getPastTrades` with the addition of an optional `fromBlock` param. To know when to call though, we'd subscribe to bloom filters.
Is that it?

Reply:
> What I can tell is that, while testing, I'm somehow subscribing more than once to the events, or at least getting more than one notification per event. Tried to always unsubscribe before new subscriptions but that still did not help.

that's strange, probably we do more than one subscription? maybe is the hook, that allows to reuse the logic, but if you for example do a subscription in a hook, and this hook is being used in 2 components, you end up with 2 subscriptions (of course, you can handle it with some logic)

> In summary, you are saying that we might just not have this method, and instead rely on getPastTrades with the addition of an optional fromBlock param. To know when to call though, we'd subscribe to bloom filters.
Is that it?

As a wise man once said, "I can't testify in favor nor against any approach.". But yes, I was suggesting that my gut was telling me less problems and more control if we use `getPastEvents` 😗

Reply:
The last version is using only `getPastEvents`, still have to add the polling logic though.
It think for now (while I write the other parts) I'll do a simple interval.
Later we can improve with the bloom filter approach.